src/sink.rs

```diff
@@ -0,0 +1,277 @@
+    /// The caller is responsible for ensuring all bytes up to `new_pos` are properly initialized.
+    #[cfg(not(all(feature = "safe-encode", feature = "safe-decode")))]
+    #[inline]
+    pub unsafe fn set_pos(&mut self, new_pos: usize) {
+        debug_assert!(new_pos <= self.capacity());
+        self.pos = new_pos;
+    }
+
+    /// Returns the initialized section of the Sink. Analogous to `&[..pos]`.
+    #[cfg(any(test, feature = "safe-decode"))]
+    #[inline]
+    pub fn filled_slice(&self) -> &[u8] {
+        // SAFETY: from the safety invariant.
+        unsafe { core::slice::from_raw_parts_mut(self.output.as_ptr() as *mut u8, self.pos) }
+    }
+}
+
+/// A Sink wrapper backed by a Vec<u8>.
+pub struct VecSink<'a> {
+    sink: Sink<'a>,
+    offset: usize,
+    vec_ptr: *mut Vec<u8>,
+}
+
+impl<'a> VecSink<'a> {
+    /// Creates a `Sink` backed by the Vec bytes at `vec[offset..vec.capacity()]`.
+    /// Note that the bytes at `vec[output.len()..]` are actually uninitialized and will
+    /// not be readable until written.
+    /// When the `Sink` is dropped the Vec len will be adjusted to `offset` + `Sink.pos`.
+    #[inline]
+    pub fn new(output: &'a mut Vec<u8>, offset: usize, pos: usize) -> Self {
+        // SAFETY: Only the first `output.len` in `output` are initialized.
+        // Assert that the range output[offset + pos] is all initialized data.
+        let _ = &output[..offset + pos];
+
+        // SAFETY: Derive the pointer first, for stacked borrows reasons.
+        let vec_ptr = output as *mut Vec<u8>;
+
+        // SAFETY: `Vec` guarantees that `capacity` elements are available from `as_mut_ptr`.
+        // Only the first `output.len` elements are actually initialized but we use a slice of `MaybeUninit` for the entire range.
+        // `MaybeUninit<T>` is guaranteed to have the same layout as `T`.
+        let vec_with_spare = unsafe {
+            core::slice::from_raw_parts_mut(
+                output.as_mut_ptr() as *mut MaybeUninit<u8>,
+                output.capacity(),
+            )
+        };
+        VecSink {
+            sink: Sink {
+                output: &mut vec_with_spare[offset..],
+                pos,
+            },
+            offset,
+            vec_ptr,
+        }
+    }
+}
+
+impl<'a> Deref for VecSink<'a> {
+    type Target = Sink<'a>;
+    #[inline]
+    fn deref(&self) -> &Self::Target {
+        &self.sink
+    }
+}
+
+impl<'a> DerefMut for VecSink<'a> {
+    #[inline]
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.sink
+    }
+}
+
+impl<'a> Drop for VecSink<'a> {
+    #[inline]
+    fn drop(&mut self) {
+        unsafe { (&mut *self.vec_ptr).set_len(self.offset + self.sink.pos) }
+    }
+}
+
+#[inline]
+fn slice_as_uninit_ref(slice: &[u8]) -> &[MaybeUninit<u8>] {
+    // SAFETY: `&[T]` is guaranteed to have the same layout as `&[MaybeUninit<T>]`
+    unsafe { core::slice::from_raw_parts(slice.as_ptr() as *mut MaybeUninit<u8>, slice.len()) }
+}
```

Comment:
it's weird that using MaybeUninit requires to add more unsafe code

Reply:
Right?! It's baffling sometimes.